feat: add experiment detail actions [DET-3083]

hamidzr opened the pull request:
## Description


depends on the tensorboard launch action PR
## Test Plan



## Commentary (optional)


<!-- User-facing API changes need the "User-facing API Change" label -->

<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")

## Description

Lead with the intended commit body in this description field. For breaking changes, please include "BREAKING CHANGE:" at the beginning of your commit body. At minimum, this section should include a bracketed reference to the Jira ticket, e.g. "[DET-1234]". When squash-and-merging, copy this directly into the description field.

## Test Plan

Describe the situations in which you've tested your change, and/or a screenshot as appropriate. Reviewers may ask questions of this test plan to ensure adequate manual coverage of changes.

## Commentary (optional)

Use this section of your description to add context to the PR. Could be for particularly tricky bits of code that could use extra scrutiny, historical context useful for reviewers, etc. You may intentionally leave this section blank and remove the title.
--->


[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

hamidzr commented on webui/react/src/hooks/useRestApi.ts:
```diff
@@ -45,8 +45,11 @@ type Output<T> = [
-    case ActionType.SetData:
-      return { ...state, data: action.value, hasLoaded: true, isLoading: false };
+    case ActionType.SetData: {
+      // TODO unnecessary?
```
this isn't necessary for this PR but @hkang1 do you think this is necessary in general or react is doing something similar under the hood?

hkang1 replied:
I wasn't sure either, so I looked into it. Looks like react uses the native JS `Object.is` function to [compare objects](https://reactjs.org/docs/hooks-reference.html#bailing-out-of-a-dispatch). According to the [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is#Description) it only compares to see if the references are the same for the objects. So in this example, `state.data` is always going to be a different reference than `action.value` (unless you are actually passing in the previous value of `action.value`). But looking at this, it looks like are want to check to make sure the shape and content of the objects are identical so this is NOT redundant code.

hkang1 replied:
TL;DR This is great to have :)

hamidzr commented on webui/react/src/components/ExperimentActions.tsx:
```diff
@@ -0,0 +1,161 @@
+
+interface Props {
+  experiment: ExperimentDetails;
+  updateFn: () => void;
```
rename to `onSuccess` or `onDone`

hamidzr commented:
updated this after #836

hkang1 commented on webui/react/src/components/ExperimentActions.tsx:
```diff
@@ -0,0 +1,169 @@
+
+interface Props {
+  experiment: ExperimentDetails;
+  finally: () => void; // A callback to trigger after an action is done.
```
**nit:** on callbacks I've been using the format of `on[Action]` so for this maybe we can name it `onComplete`.

hamidzr replied:
sg

hamidzr replied:
with `finally` I'm trying to borrow from the async api pattern and to communicate that this will get called wether the action successfully completes or it fails.

hamidzr replied:
`onSetteled` ? after https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled

hkang1 replied:
`onSettled` sounds good to me!

hkang1 commented on webui/react/src/components/ExperimentActions.tsx:
```diff
@@ -0,0 +1,169 @@
+
+const ExperimentActions: React.FC<Props> = ({ experiment, finally: updateFn }: Props) => {
+
+  interface ButtonLoadingStates {
+    kill: boolean;
+    [RunState.StoppingCanceled]: boolean;
+    archive: boolean;
+    [RunState.Paused]: boolean;
+    [RunState.Active]: boolean;
+    tsb: boolean; // tensorboard
+  }
```
**chore:** move this below `Props` definition

hkang1 commented on webui/react/src/components/ExperimentActions.tsx:
```diff
@@ -0,0 +1,169 @@
+    tsb: false,
+  });
+
+  const killExperimentCB = useCallback(() => {
```
**nit:** rename to `handleKill`

hkang1 commented on webui/react/src/components/ExperimentActions.tsx:
```diff
@@ -0,0 +1,169 @@
+      .finally(() => setButtonStates(state => ({ ...state, kill: false })));
+  }, [ experiment.id, updateFn ]);
+
+  const archiveCB = useCallback((archive: boolean) =>
```
**nit:** `handleArchive` and maybe move it above `handleKill` for alphabetical ordering?

hkang1 commented on webui/react/src/components/ExperimentActions.tsx:
```diff
@@ -0,0 +1,169 @@
+const ExperimentActions: React.FC<Props> = ({ experiment, finally: updateFn }: Props) => {
+
+  interface ButtonLoadingStates {
+    kill: boolean;
+    [RunState.StoppingCanceled]: boolean;
+    archive: boolean;
+    [RunState.Paused]: boolean;
+    [RunState.Active]: boolean;
+    tsb: boolean; // tensorboard
```
**suggestion:** Create a separate enum just for experiment actions. In this [PR](https://github.com/determined-ai/determined/pull/866/files) I defined something like this:
```
enum Action {
  Activate = 'Activate',
  Archive = 'Archive',
  Cancel = 'Cancel',
  Kill = 'Kill',
  Pause = 'Pause',
  OpenTensorBoard = 'Open Tensorboard',
}
```

hamidzr replied:
this is a lot better :+1:

hamidzr replied:
how does this look? 

```
enum Action {
  Archive = 'Archive',
  Kill = 'Kill',
  Tensorboard = 'Tensorboard',
}

type ButtonLoadingStates = Record<Action | RunState, boolean>;
```

hkang1 commented on webui/react/src/components/ExperimentActions.tsx:
```diff
@@ -0,0 +1,169 @@
+const ExperimentActions: React.FC<Props> = ({ experiment, finally: updateFn }: Props) => {
+
+  interface ButtonLoadingStates {
+    kill: boolean;
+    [RunState.StoppingCanceled]: boolean;
+    archive: boolean;
+    [RunState.Paused]: boolean;
+    [RunState.Active]: boolean;
+    tsb: boolean; // tensorboard
```
**nit:** `tensorboard` instead of `tsb`

hkang1 commented on webui/react/src/components/ExperimentActions.tsx:
```diff
@@ -0,0 +1,169 @@
+
+  return (
+    <ul className={css.base}>
+      {actionButtons
+        .filter(ab => !ab.showIf || ab.showIf(experiment as ExperimentDetails))
+        .map(ab => ab.btn)
+      }
```
**suggestion:** We talk more about this one async if you want. In short, this might have too many layers of abstraction. IMO just having all the JSX buttons listed out here with all the properties set will keep this file shorter and easier to read. For an example, check out this [PR](https://github.com/determined-ai/determined/pull/866/files), specifically `ExperimentList.tsx` lines 224-245. With the current setup, if you need to edit the `Cancel` button click handler you have to find `actionButtons` then look for `cancelButton` then finally `requestExpStateCB`.

hamidzr replied:
hmm let's chat about this.

hkang1 replied:
We discussed this async and decided the array filter and map still makes sense but decided to move the buttons inline into the `actionButtons` list to reduce vertical file space.

hkang1 commented on webui/react/src/utils/data.ts:
```diff
@@ -87,3 +87,8 @@ export const taskStateSorter = (a: State, b: State): number => {
   return aValue - bValue;
 };
+
+export const isEqual = (a: unknown, b: unknown): boolean => {
+  if (a === b) return true;
+  return JSON.stringify(a) === JSON.stringify(b);
+};
```
nice, this is sweet!

hamidzr replied:
: ) would `areEqual` make more sense?

hkang1 commented on webui/react/src/ioTypes.ts:
```diff
@@ -136,6 +136,7 @@ const validationHistoryIoType = io.type({
 const trialSummaryIoType = io.type({
   hparams: io.any,
   id: io.number,
+  num_steps: io.number,
```
**suggestion:** you can leave this to a separate PR but there is a `num_batches: io.number` available now.

hamidzr replied:
I added it in.

hkang1 commented:
**nit:** this can be left for a separate PR
![2020-07-14 at 9 59 AM](https://user-images.githubusercontent.com/220971/87449177-11b6c100-c5ba-11ea-9add-cec44dcce3a4.png)

hamidzr commented:
thanks for the review I addressed the comments there are two left to chat about

https://github.com/determined-ai/determined/pull/858#discussion_r454448836
and https://github.com/determined-ai/determined/pull/858#issuecomment-658269574

hamidzr commented:
> **nit:** this can be left for a separate PR
> ![2020-07-14 at 9 59 AM](https://user-images.githubusercontent.com/220971/87449177-11b6c100-c5ba-11ea-9add-cec44dcce3a4.png)

IMO having the kill and cancel buttons solid but the action buttons as outlines looks a bit off. I can make all outlines to match batchations :thinking: .. I'll just keep them as primary till we have the rest of the page in place and see what works best?

hamidzr commented:
rebased to resolve conflicts with master.
